Follow up to PR #418

SHuang-Broad opened the pull request:
Following up #418 with more optimizations, and two new pipelines for 

ingesting single demultiplexed PacBio Hifi (u)BAM (aka a readgroup uBAM)

SHuang-Broad committed:
UPDATES TO THE Hifiasm pipeline:
  * update Hifiasm to version 0.19.5
  * update how Hifiasm outputs are compressed (bgz replacing gz), also
  * monitor hifiasm resources usage

SHuang-Broad committed:
For both CCS/ONT, update PBSV
  * update docker used in PBSV tasks to the version coming with official SMRTLink releases (2.9.0)
  * change how the 2-step PBSV process is done (following the recommended way now)

SHuang-Broad committed:
For both CCS/ONT, update Sniffles-2
  * to version 2.0.7
  * using TRF bed
  * conditionally phase sv (requires phased bam)
  * generates its own vcf.gz and tbi

SHuang-Broad committed:
MAJOR REFACTOR: UNIFY CCS/ONT WGS PIPELINE

Overhaul how small variants are called in the WG pipelines

  * default to use DV to call small variants, Clair3 analysis needs to be requested explicitly
  * retire the Pepper toolchain completely from the CCS pipeline, using DV directly
  * for R10.4+ ONT data, also use DV directly
  * older ONT data would still use the PEPPER-DV-Margin pipeline
  * offers GPU version (though based on, it's not worth it yet)
  * update how bam haplotagging is done

Cleanup structural variants calling
  * experiment with SNF2 phasing SV calls (implicitly depends on small variants calling now)
  * tune PBSV calling
    - discover now supports --hifi
    - output vcf.gz and tbi
    - less verbose logging by default

Misc.:
  * optimizations to BAM merging and metrics workflow
  * updates coverage collection step
  * new R script to visualize log from vm_monitoring_script.sh

SHuang-Broad committed:
CRITICAL: updating pbindex to that from the SMRTLink 12 release

SHuang-Broad committed:
New docker that's intended to replace lr-basic:
  * incorporates gcloud cli (not just gsutil)
  * integrate libdeflate for more speedups

tedsharpe commented on wdl/tasks/Utility/BAMutils.wdl:
```diff
@@ -43,3 +52,64 @@ task GetReadGroupInfo {
     }
 }
+
+task ResetSamplename {
```
This seems like an awfully dangerous thing to do (arbitrary replacement of sample name).  The task doesn't appear to be in use, currently.  I hope we'll use it judiciously.

tedsharpe commented on wdl/pipelines/PacBio/Utility/ProcessOnInstrumentDemuxedChunk.wdl:
```diff
@@ -0,0 +1,98 @@
+    ###################################################################################
+    # generate PBI
+    call PBUtils.PBIndex as Index {input: bam = uBAM}
+    call BU.GetReadGroupInfo as RG {input: uBAM = uBAM, keys = ['SM', 'LB', 'PU']}
```
doesn't look like we actually need the SM value

tedsharpe reviewed (approved):
Just a few suggestions.  Might want to omit "/cromwell_root" wherever it appears, as I suggested in the last PR.

SHuang-Broad committed:
MISC UPDATES TO SEVERAL UTILS TASKS

  * organize dockstore.yml file a bit

  * make WDL validation shell script more usable

  * update pbmm2 and pbindex to versions in SMRTLink

  * update GeneralUtils.wdl
    - two bash-like new tasks [CoerceMapToArrayOfPairs, CoerceArrayOfPairsToMap]
    - cleanup task CollapseArrayOfStrings

  * update resource allocations to tasks
    - NanoplotFromBam (also changes docker)
    - MosDepthWGS

SHuang-Broad committed:
significantly boost capabilities of BAMutils.wdl

incorporate new tasks and optimize them

  * [CountMethylCallReads, GatherReadsWithoutMethylCalls]
    from sh_beans

  * [GetPileup, BamToRelevantPileup]
    from sh_more_atomic_qc

  * [GetReadGroupLines, GetSortOrder, SplitNameSortedUbam]
    from sh_ont_fc

  * [SamtoolsFlagStats, ParseFlagStatsJson]
    from sh_trvial_stats

  * [FilterBamByLen, InferSampleName]
    from sh_seqkit

  * [CountAlignmentRecords, StreamingBamErrored, CountAlignmentRecordsByFlag]
    from sh_maha_aln_metrics

  * [ResetSamplename]
    from sh_ingest_singlerg

  * [MergeBamsWithSamtools]
    from sh_ont_fc.Utils.wdl

  * [BamToFastq]
    from sh_more_bam_qcs
    and optimize it with
    sh_ingest_singlerg.Utils.wdl

delete
  * GetSortOrder as that's now implemented in GatherBamMetadata
  * Drop2304Alignments as that's no longer used

update dockers to the latest

SHuang-Broad committed:
New docker that updates the samtools in GATK to the latest (1.18)

SHuang-Broad committed:
New docker that offers a new mode remove duplicates from a queryname-sorted ONT bam

SHuang-Broad committed:
update some legacy code due to new interfaces in updated BAMutils.wdl

SHuang-Broad committed:
New workflows
CHERRY-PICK FROM VARIOUS QC/METRICS BRANCHES:

  * collect information about ML/MM tags in a long-read BAM
    (sh_beans)

  * a heuristic way to find peaks in a distribution (using dyst)
    (sh_dyst_peaker)

  * filter reads by length in a BAM
  * collect some read quality stats from (length-filtered) FASTQ/BAM
    (sh_seq_kit)

  * VerifyBamID2 (for contamination estimation)
  * naive sex-concordance check
    (sh_more_atomic_qc)

  * check fingerprint of a single BAM file
    (sh_sample_fp)

  * collect SAM flag stats
    (sh_trivial_stats)

SHuang-Broad committed:
DEPRECATION

  * AlignAndCheckFingerprintCCS.wdl
  * CollectPacBioAlignedMetrics.wdl
  * CollectSMRTCellUnalignedMetrics.wdl

SHuang-Broad committed:
Fix bug in deduplicating aligned ONT BAM
  (CHRRY-PICK & follow up to PR 406)

SHuang-Broad committed:
New workflow to save files (avoid multiple calls to Finalize)

SHuang-Broad committed:
Improve various existing codes
  * make BeanCounter finalization optional
    (wdl/pipelines/TechAgnostic/Utility/CountTheBeans.wdl)
  * custom struct for sub-workflow config using a JSON
    (wdl/pipelines/TechAgnostic/Utility/LongReadsContaminationEstimation.wdl)
  * make fingerprint checking subworkflow control size filtering
    (wdl/tasks/QC/FPCheckAoU.wdl)
    (wdl/pipelines/TechAgnostic/Utility/VerifyBamFingerprint.wdl)
  * fix a warning by IDE/miniwdl complaining WDL stdlib function length only applies to Array
    (wdl/tasks/Utility/BAMutils.wdl)
  * various updates to Finalize
    (wdl/tasks/Utility/Finalize.wdl)

New tasks in (wdl/tasks/Utility/GeneralUtils.wdl) to
  * correctly convert Map to TSV
  * concatenate files

SHuang-Broad committed:
New workflow to unify QC and metrics collection on aligned BAM

SHuang-Broad committed:
Update grouping of pipelines in .dockstore.yml

SHuang-Broad committed:
update WDL validation bash script

SHuang-Broad committed:
New subworkflow to align PacBio uBAM

SHuang-Broad committed:
Two new workflows:
    handle (ref-indepedent, alignment) a single piece of on-instrument demultiplexed CCS bam

SHuang-Broad committed:
DEPRECATION:
  * SampleLevelAlignedMetrics.wdl
  * PBCLRWholeGenome.wdl

SHuang-Broad committed:
Update utility code:
  * new struct in AlignedBamQCandMetrics.wdl to facilicate as-sub-workflow calling
  * change parameters name for fingerprint workflows

SHuang-Broad committed:
a few tweaks to to AlignedBamQCandMetrics:
  * make saving of reads without methylation SAM tags optional
  * better parameter naming

SHuang-Broad committed:
correct two mistakes: 1. a typo; 2. should append not replace

SHuang-Broad committed:
Update Pacbio alignment task

SHuang-Broad committed:
make read-length util tasks pre-emptible

SHuang-Broad committed:
make pbindex task more frugal

SHuang-Broad committed:
make fingerprint reads extraction more efficient/resilient

SHuang-Broad committed:
fix a stupid logical typo error (in saving files)

SHuang-Broad committed:
option to force run fingerprintcheck on small bams

SHuang-Broad committed:
Safer and more efficient way to do targetted pileup conversion
(affects contamination estimation)

SHuang-Broad committed:
bug fix for bam2fastq

SHuang-Broad committed:
optimize alignment further

SHuang-Broad committed:
test reducing resources for pbmm2 further

SHuang-Broad committed:
some more resource optimization

SHuang-Broad committed:
update PacBio post-alignment QC optionality

SHuang-Broad committed:
now we can rely on Terra itself for fast localization in PBUtils.wdl

SHuang-Broad committed:
add output tracking how long the alignment step took
